dali/kernels/signal/resampling_gpu.cuh

```diff
@@ -0,0 +1,138 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+#define DALI_KERNELS_SIGNAL_RESAMPLING_GPU_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/signal/resampling.h"
+#include "dali/core/util.h"
+
+#define SHM_NCHANNELS 16
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+namespace resampling {
+
+struct SampleDesc {
+  void *out;
+  const void *in;
+  ResamplingWindow window;
+  int64_t in_len;  // num samples in input
+  int64_t out_begin;  // output region-of-interest start
+  int64_t out_end;  // output region-of-interest end
+  int nchannels;  // number of channels
+  double scale;  // in_sampling_rate / out_sampling_rate
+};
+
+/**
+ * @brief Gets intermediate floating point representation depending on the input/output types
+ */
+template <typename Out, typename In>
+__device__ float ConvertInput(In in_val) {
+  if (std::is_unsigned<Out>::value && std::is_signed<In>::value) {
+    return (ConvertSatNorm<float>(in_val) + 1.0f) * 0.5f;
+  } else if (std::is_signed<Out>::value && std::is_unsigned<In>::value) {
+    return ConvertSatNorm<float>(in_val) * 2.0f - 1.0f;  // treat half-range as 0
+  } else {
+    return ConvertSatNorm<float>(in_val);  // just normalize
+  }
+}
+
+/**
+ * @brief Resamples 1D signal (single or multi-channel), optionally converting to a different data type.
+ *
+ * @param samples sample descriptors
+ */
+template <typename Out, typename In, bool SingleChannel = false>
+__global__ void ResampleGPUKernel(const SampleDesc *samples) {
+  auto sample = samples[blockIdx.y];
+  double scale = sample.scale;
+  float fscale = scale;
+  int nchannels = SingleChannel ? 1 : sample.nchannels;
+  auto& window = sample.window;
+
+  extern __shared__ float sh_mem[];
+  float *window_coeffs_sh = sh_mem;
+  float *tmp = sh_mem + window.lookup_size +
+               threadIdx.x * (SHM_NCHANNELS+1);  // used to accummulate per-channel out values
+  for (int k = threadIdx.x; k < window.lookup_size; k += blockDim.x) {
+    window_coeffs_sh[k] = window.lookup[k];
+  }
+  __syncthreads();
+  window.lookup = window_coeffs_sh;
+
+  Out* out = reinterpret_cast<Out*>(sample.out);
+  const In* in = reinterpret_cast<const In*>(sample.in);
+
+  int64_t grid_stride = static_cast<int64_t>(gridDim.x) * blockDim.x;
+  int64_t out_block_start = sample.out_begin + static_cast<int64_t>(blockIdx.x) * blockDim.x;
+
+  for (int64_t out_pos = out_block_start + threadIdx.x; out_pos < sample.out_end;
+       out_block_start += grid_stride, out_pos += grid_stride) {
+    double in_block_f = out_block_start * scale;
+    int64_t in_block_i = std::floor(in_block_f);
+    float in_pos_start = in_block_f - in_block_i;
+    const In* in_blk_ptr = in + in_block_i * nchannels;
+    float in_pos = in_pos_start + fscale * threadIdx.x;
```

Comment:
I would be in favor of adding the explanation of the block/thread setup as a comment here, since Michał already have done a nice visualization of it in the review comment. Both how it is and why it is this way as the why fact is rather obscure in my opinion.

Reply:
Actually we ended up with the original layout, but calculating `in_block_f` inside the loop, since the precision problem could be noticeable with either layout. I'll try to add an explanation on what's going on here in the comments